android-libs/src/main/cpp/reporter.cpp

```diff
@@ -0,0 +1,106 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "mkall.h"
+
+#include <string>
+#include <vector>
+
+#include "vendor/json.hpp"
+#include "vendor/mkcollector.hpp"
+
+#include "mkall_util.h"
+
+JNIEXPORT jlong JNICALL Java_io_ooni_mk_MKReporterTask_New(
+    JNIEnv *env, jclass /*clazz*/, jstring softwareName, jstring softwareVersion) {
+  if (env == nullptr || softwareName == nullptr || softwareVersion == nullptr) {
+    MKALL_THROW_EINVAL;
+    return 0L;
+  }
+  std::string software_name;
+  std::string software_version;
+  if (!mkall_string_java2cxx(env, softwareName, software_name) ||
+      !mkall_string_java2cxx(env, softwareVersion, software_version)) {
+    return 0L;  // Exception already pending
+  }
+  return reinterpret_cast<jlong>(new mk::collector::Reporter{
+      std::move(software_name), std::move(software_version)});
+}
+
+MKALL_SET_STRING(
+    MKReporterTask_SetCABundlePath,
+    [](mk::collector::Reporter *r, const char *s) noexcept {
+      r->set_ca_bundle_path(s);
+    },
+    mk::collector::Reporter)
+
+// mk_reporter_results contains the results of Submit
+struct mk_reporter_results {
+  bool good = false;
+  std::vector<std::string> logs;
+  std::string measurement;
+  std::string report_id;
+  std::string stats;
+};
+
+JNIEXPORT jlong JNICALL Java_io_ooni_mk_MKReporterTask_Submit(
+    JNIEnv *env, jclass /*clazz*/, jlong handle, jstring measurement, jlong uploadTimeout) {
+  if (env == nullptr || handle == 0L || measurement == nullptr) {
+    MKALL_THROW_EINVAL;
+    return 0L;
+  }
+  auto results = new mk_reporter_results{};
+  if (!mkall_string_java2cxx(env, measurement, results->measurement)) {
+    return 0L;  // Exception already pending
+  }
+  mk::collector::Reporter::Stats stats;
+  auto r = reinterpret_cast<mk::collector::Reporter *>(handle);
+  results->good = r->submit_with_stats(
+      results->measurement, results->logs, uploadTimeout, stats);
+  results->report_id = r->report_id();
+  {
+    nlohmann::json doc;
+#define XX(name_) doc[#name_] = stats.name_;
+    MKCOLLECTOR_REPORTER_STATS_ENUM(XX)
+#undef XX
+    results->stats = doc.dump();
+  }
+  return reinterpret_cast<jlong>(results);
```

Comment:
Justify why using `reinterpret_cast<>` here.

Reply:
According to [cppreference,com](https://en.cppreference.com/w/cpp/language/reinterpret_cast):

> [...] the reinterpret_cast expression does not compile to any CPU instructions (except when converting between integers and pointers or on obscure architectures where pointer representation depends on its type).
>
> Only the following conversions can be done with reinterpret_cast:
> [...]
> 2) A pointer can be converted to any integral type large enough to hold all values of its type [...]
>
> 3) A value of any integral or enumeration type can be converted to a pointer type. A pointer converted to an integer of sufficient size and back to the same pointer type is guaranteed to have its original value [...]

So, here we're casting with reinterpret cast the pointer to `jlong` (case 2). Then, when we are to use the pointer, we cast back (case 3). We're good.